### PR TITLE
Implementation of min0() using IntrinsicFunction 

### PR DESCRIPTION
**Pull request by HarshitaKalani**
Implemented min0

### CODE REVIEW

**Comment by HarshitaKalani** [comment]
One of the checks failed here but I'm not sure of the error. Please suggest.

**Comment by certik** [comment]
The error is:
```
/usr/bin/ld: ../src/libminpack.a(qrfac.f.o): in function `_lcompilers_min0_i32':
LFortran:(.text+0x0): multiple definition of `_lcompilers_min0_i32'; ../src/libminpack.a(hybrd.f.o):LFortran:(.text+0x0): first defined here
/usr/bin/ld: ../src/libminpack.a(qform.f.o): in function `_lcompilers_min0_i32':
LFortran:(.text+0x0): multiple definition of `_lcompilers_min0_i32'; ../src/libminpack.a(hybrd.f.o):LFortran:(.text+0x0): first defined here
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
The issue is that now we are creating the automatically generated function `_lcompilers_min0_i32` in two different separate files that for Minpack are compiled to the object binary file separately, and then they collide at link time (multiple definitions). To solve this, I can see two approaches:


* One approach is using some unique prefix/postfix for these functions; how to make this unique for the separate compilation that Minpack uses? I think LFortran could provide a unique hash that is guaranteed to be different for each invocation of LFortran, the main driver computes it and makes it available. Then the pass simply adds this hash to the name of these automatically generated functions. How to obtain a unique hash? We can start by generating some pseudo random number and create a long enough hash. We can investigate some options, I think this might work in practice, and in the extremely unlikely case that it would collide, simply rerunning the compilation would fix it. It's not ideal, but might end up extremely robust in practice. Just like git hashes in principle are not unique, but in practice they are.

* The other approach is to somehow store these "instantiations" separately and then reuse them / link them properly later. This approach is also fragile, since now we need to either modify the build system, or hope that the next invocation of LFortran finds these function saved in some separate file, etc. I think this is very fragile.

Why don't we implement the first option? That still leaves the door to implement the second if needed later, but I suspect the first approach can be done very robustly.

**Comment by Pranavchiku** [comment]
The first approach can be implemented in this way:
We create an intrinsic function say it `f`. Let's get the hash of this function and append it in `f->m_name` that way it will be unique.

**Comment by certik** [comment]
I think our hashes are unique for a given LFortran run, but not across independent runs. However I think the solution is UUID (https://en.wikipedia.org/wiki/Universally_unique_identifier) which should be globally unique, across invocation of LFortran.

**Comment by certik** [comment]
Let's use this function:
```c++
std::string get_unique_ID() {
    static std::random_device dev;
    static std::mt19937 rng(dev());
    std::uniform_int_distribution<int> dist(0, 61);
    const std::string v =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
    std::string res;
    for (int i = 0; i < 22; i++) {
        res += v[dist(rng)];
    }
    return res;
}
```
This creates base62 number of 22 characters, roughly the same randomness as UUID. UUID is 128 bits, this function is log(62)/log(2) * 22 = 130.99 bits, so it's tiny bit more, so we should be fine.

**Comment by certik** [comment]
@HarshitaKalani, in here: https://github.com/lfortran/lfortran/blob/addc5200215ef65c45407edbf6b50b41b26f71d9/src/libasr/asr_scopes.cpp#L84, can you please call the above `get_unique_ID` function and append the result something like this:
```diff
-    std::string unique_name = name;
+    std::string unique_name = name + lcompilers_unique_ID;
```

Then declare a global variable `lcompilers_unique_ID` and set it in the main program, and set it to `get_unique_ID` if `--generate-object-code` is set, otherwise to "", so that our ASR and LLVM tests still pass (otherwise every unique name would be truly unique and change with every run).

We can later refactor this more, but that should be good enough for now.

**Comment by HarshitaKalani** [comment]
> Then declare a global variable lcompilers_unique_ID and set it in the main program,

So I defined the lcompilers_unique_ID globally. By `setting in the main program`, do you mean writing an int main(){ } and the if-else statement inside it?
If so, I'm getting this error
```
error: use of undeclared identifier 'compiler_options'
extern std::string lcompilers_unique_ID  = compiler_options.generate_object_code ? get_unique_ID() : "";
```

**Comment by certik** [comment]
Yes, like this. Search for compiler_options in the driver, you should be able to find it. If not, I can have a look later.

**Comment by HarshitaKalani** [comment]
For defining compilers_unique_ID globally, I wrote `std::string lcompilers_unique_ID;` in lfortran.cpp file. Is this the right way of doing this?

File: src/bin/lfortran.cpp [hunk 1]
```diff
@@ -919,6 +934,8 @@ int compile_to_object_file(const std::string &infile,
         e.save_object_file(*(m->m_m), outfile);
     }
 
+    lcompilers_unique_ID = compiler_options.generate_object_code ? get_unique_ID() : "";
```
Comment (certik) [inline src/bin/lfortran.cpp:937]: I would move this to the main program at the top somewhere.

File: src/libasr/asr_scopes.cpp [hunk 2]
```diff
@@ -6,6 +6,8 @@
 
 namespace LCompilers  {
 
+std::string lcompilers_unique_ID;
```
Comment (certik) [inline src/libasr/asr_scopes.cpp:9]: In here I think you have to put "extern" to make it the same global variable as in the main program.

**Comment by certik** [comment]
Yes, I think this is exactly how to do it. I left some comments. You can test it by using the `--generate-object-code` option to make sure it works.

File: src/libasr/asr_scopes.cpp (hunk 2 above)
Comment (HarshitaKalani) [reply src/libasr/asr_scopes.cpp:9]: On using 'extern' here, I'm getting the following error:
```
Undefined symbols for architecture x86_64:
  "LCompilers::lcompilers_unique_ID", referenced from:
      LCompilers::SymbolTable::get_unique_name(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) in libasr.a(asr_scopes.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Comment (certik) [reply src/libasr/asr_scopes.cpp:9]: I think that's when you link it without the driver. Let's switch it --- put the extern into the main program, and no extern here. Then it should work.

**Comment by Pranavchiku** [comment]
Extern has to be put in `asr_scopes`, we have to define variables outside `namespaces` and put the assignment to `lcompilers_unique_ID` at correct place.

**Comment by Pranavchiku** [comment]
I checked locally, minpack passes. Let's wait for CI to run.

**Review by Pranavchiku — APPROVED** [review]
LGTM! Thank you doing this :)

File: src/libasr/asr_scopes.cpp [hunk 3]
```diff
@@ -81,7 +83,8 @@ ASR::symbol_t *SymbolTable::find_scoped_symbol(const std::string &name,
 }
 
 std::string SymbolTable::get_unique_name(const std::string &name) {
-    std::string unique_name = name;
+    std::cout<<"lcompilers_unique_ID: "<<lcompilers_unique_ID<<"\n";
```
Comment (certik) [inline src/libasr/asr_scopes.cpp:86]: Remove this line.

File: src/bin/lfortran.cpp [hunk 4]
```diff
@@ -2080,7 +2097,7 @@ int main(int argc, char *argv[])
                 throw LCompilers::LCompilersException("Unsupported backend.");
             }
         }
-
+
```
Comment (certik) [inline src/bin/lfortran.cpp:2100]: Remove white space.

**Review by certik — APPROVED** [review]
I think that this looks great now!

Thanks @HarshitaKalani and @Pranavchiku.

**Comment by HarshitaKalani** [comment]
Thanks @Pranavchiku !

**Comment by certik** [comment]
I tested the latest main by hand as well, it seems to work well. The `--generate-object-code` option enables the unique name and the name is unique across compiler invocations, so every file in Minpack is compiled with its own unique instantiations, and so things do not collide.

**Pull request merged**